Update waiter with login prompt if not authorized

karawoo opened the pull request:
I think this fixes #17 for the reticulate branch. If approved I can make the same change to the synapser version as well. With this change, the waiter screen updates and tells the user to log in if the request to get the session token returns a 401. I've deployed a copy of this app here: https://shinypro.synapse.org/users/kwoo/SynapseShinyApp-test/ (I'll update it with any future changes and take it down when this is resolved).

I've left the same `waiter_update()` in the `tryCatch()` block in case logging in _does_ throw an error. Do we want to make this a separate function since it's now repeated twice?

karawoo committed:
Show prompt to log in if not authorized

Aryllen commented:
Looks to be behaving as it should now. When I first logged out and then tried the app, it didn't populate the message saying I wasn't logged in; instead, everything went away, including the logo, leaving only the purple background color. I only had that happen the one time, though, and it was immediately after signing out.

A separate function might be nice for the `waiter_update()`. I am not sure what that function would look like, though.

karawoo commented:
We agreed in the code review meeting that we'd update the js function so that it's not creating two separate inputs. If the request returns a 401 we can save some useful information to the cookie input. On the shiny side, in the observer we'll check the cookie value and update the waiter accordingly (and keep the try/catch block in case the person is logged in but the session token is expired).

karawoo committed:
Make readCookie pass response text to shiny regardless of response code

karawoo committed:
Show login prompt if input$cookie is an empty string

karawoo committed:
Put rest of synapse activity in else block

karawoo committed:
Save specific string to input$cookie if unauthorized

This is needed to be able to run the app locally as well as on the server. When
running locally, we want to fall back to logging in via credentials in
.synapseConfig since the session token cookie is not available. However since
it's not available, its value is "" which means we can never make it past the
first if block. With this change, when the app is running on the server and the
user isn't logged in, the shiny input gets the value "unauthorized" and prompts
the user to log in, but when running locally it gets the empty string and can
log in using the .synapseConfig

karawoo commented:
I've made the updates we talked about on Friday. @thomasyu888, in your previous update you took out the header above the plot that shows the synapse user name. I think that is actually useful to keep in the sample app because it shows where the author needs to put code that depends on synapse (i.e., it can't be outside the observer for `input$cookie`) so I also put that back with a comment. @marcomarasca if you have a chance to take a look at this I'd love your thoughts.

I've updated the example running at https://shinypro.synapse.org/users/kwoo/SynapseShinyApp-test/ so you can test it out. I'm also able to run it locally.

thomasyu888 commented:
Looks good to me!  Thanks for the contribution.

marcomarasca reviewed (approved):
LGTM!
